File: python/05_ArrayStack.py
'''
	스택: 알고리즘 구현(Python) -- 순차 자료구조
		파일명: ArrayStack.py
			- 스택 생성.소멸	: __init__, __del__
			- 스택 상태			: empty
			- 스택 크기			: size
			- 데이터 삽입.삭제	: push, pop
			- 데이터 확인(peek)	: top
			- 전체 데이터 출력	: printStack
'''

class ArrayStack: 
    # 빈 스택 생성
    def __init__(self):
        self.__stack = []

    # 스택 삭제
    def __del__(self):
        self.__stack.clear()

    # 빈 스택 여부 확인
    def empty(self) -> bool:
        if not self.__stack:
            return True
        return False

    # 스택의 원소 개수
    def size(self) -> int:
        return len(self.__stack)
                
    # 데이터 삽입: 스택에 새로운 데이터 추가
    def push(self, num) -> None:
        self.__stack.append(num)
        
    # 데이터 삭제: 스택에서 맨 위의 데이터 삭제
    def pop(self) -> None:
        self.__stack.pop()

    # 스택에서 맨 위의 데이터 반환(peek)
    def top(self):
        if not self.__stack:
            return None
        return self.__stack[-1]        

    # 스택의 전체 데이터 출력
    def printStack(self) -> None:
        if not self.__stack:
            print('입력된 데이터가 없습니다!!!')

        print('\n\t##### 입력된 데이터 #####')
        print('STACK [', end = ' ')
        for i in range(len(self.__stack)-1, -1, -1):
            print(self.__stack[i], end = ' ')
        print(']')

if __name__ == '__main__':
    import os		# system
    import sys  	# exit

    s = ArrayStack()
    while True:
        os.system('cls')
        print('\n ### 스택 구현: 1차원 배열 ###')
        print('1) 데이터 삽입: push')
        print('2) 데이터 삭제: pop')
        print('3) 전체 출력')
        print('4) 프로그램 종료\n')
        print('메뉴 선택: ', end='')
        choice = int(input())

        match choice:
            case 1:
                while True:
                    num = int(input('삽입 할 데이터 입력(종료: 0): '))
                    if num == 0:
                        break
                    s.push(num)
            case 2:
                print(f'\n삭제 된 데이터: {s.top()}')
                s.pop()
            case 3:
                s.printStack()
            case 4:
                sys.exit("\n프로그램 종료!!!")
            case _: print('\n잘못 선택 하셨습니다. \n')
        os.system('pause')

    # del s
    # s.__del__

File: cpp/05_ArrayStack.cpp
/*
	스택: 알고리즘 구현(CPP) -- 순차 자료구조
		파일명: ArrayStack.cpp
			- 스택 생성.소멸	: ArrayStack, ~ArrayStack
			- 스택 상태			: empty, full
			- 스택 크기			: size
			- 데이터 삽입.삭제	: push, pop
			- 데이터 확인(peek)	: top
			- 전체 데이터 출력	: printStack
*/

#include <iostream>
using namespace std;

// ArrayStack class
template <typename T>
class ArrayStack {
public:
	ArrayStack(int size = 10);				// 빈 스택 생성
	~ArrayStack(void);						// 스택 삭제
	bool	empty(void) const;				// 빈 스택 여부 확인
	bool	full(void) const;				// 스택의 포화 상태 여부 확인
	int		size(void) const;				// 스택의 원소 개수
	void	push(const T& data);			// 데이터 삽입
	void	pop(void);						// 데이터 삭제
	T		top(void) const;				// 스택에서 맨 위의 데이터 확인(peek)
	void	printStack(void) const;			// 스택의 전체 데이터 출력
private:
	int		top_;
	int		maxSize_;
	T		*stack_;
};

// 빈 스택 생성
template <typename T>
ArrayStack<T>::ArrayStack(int size)
	: top_(-1), maxSize_(size) {
	stack_ = new T[maxSize_];
}

// 스택 삭제
template <typename T>
ArrayStack<T>::~ArrayStack(void) {
	delete[] stack_;
}

// 빈 스택 여부 확인
template <typename T>
bool	ArrayStack<T>::empty(void) const {
	return top_ == -1;
}

// 스택의 포화 상태 여부 확인
template <typename T>
bool	ArrayStack<T>::full(void) const {
	return top_ == maxSize_ - 1;
}

// 스택의 원소 개수
template <typename T>
int		ArrayStack<T>::size(void) const {
	return top_ + 1;
}

// 데이터 삽입: 스택에 새로운 데이터 추가
template <typename T>
void	ArrayStack<T>::push(const T &data) {
	if (full()) {
		return;
	}
	stack_[++top_] = data;
}

// 데이터 삭제: 스택에서 맨 위의 데이터 삭제
template <typename T>
void	ArrayStack<T>::pop(void) {
	if (empty()) {
		return;
	}
	--top_;
}

// 스택에서 맨 위의 데이터 반환(peek)
template <typename T>
T	ArrayStack<T>::top(void) const {
	if (empty()) {
		return T();
	}
	return stack_[top_];
}

// 스택의 전체 데이터 출력
template <typename T>
void	ArrayStack<T>::printStack(void) const {
	if (empty()) {
		cout << "입력된 데이터가 없습니다!!!" << endl;
		return;
	}

	cout << "\n\t##### 입력된 데이터 #####\n" << endl;
	cout << "STACK [";
	for (int i = top_; i >= 0; --i) {
		cout.width(3);
		cout << stack_[i];
	}
	cout << " ]" << endl;
}
